### ISSUE TITLE
Generator created inside ReproducibleDataLoader leads to creating same RNG state for same worker id in each rank

### ISSUE THREAD

**Issue by gokulavasan** [pytorch/torchtune#177]
In ReproducibleDataLoader, the dataloader_seed was used create a generator which was passed into DataLoader init. This will create different RNG state for each worker for that rank (using rand(base_seed) + worker_id). But it will be the same RNG state in each worker id in each rank. That is, worker id 3 in rank0 and worker id 3 in rank1 will have the same RNG state. This is a problem.

Note that this is not an issue in torch core because there is no explicit control on determinism of transforms. That is, the torch.initial_seed for each trainer process will be different and thus each worker id will have different RNG state. When we tried to control the RNG state, we ended up making it too restrictive which is incorrect. RNG state needs to be different in each worker id across world size.

In order to fix this, the seed for each rank should be different yet deterministic - seed(base_seed + rank_id). Note that the worker id seed is created using (random(seed) + worker_id) AND not (seed + worker_id). The latter would have been problematic as seeds could easily overlap.
